File: blockchain/web3_client.py
"""
Web3 client for Save the Castle blockchain integration
Handles GameEconomy and Leaderboard contract interactions
"""

import json
from decimal import Decimal
from typing import Optional, Dict, List, Tuple
from web3 import Web3
from web3.contract import Contract
from eth_account import Account
import os
from datetime import datetime

class SaveTheCastleWeb3Client:
    """Web3 client for interacting with Save the Castle smart contracts"""
    
    # Base Sepolia Testnet configuration
    BASE_SEPOLIA_RPC = "https://sepolia.base.org"
    CHAIN_ID = 84532
    
    # Contract addresses on Base Sepolia
    GAME_ECONOMY_ADDRESS = "0x55cbea71ad8b981b91b137116b76a4828f90c548"
    LEADERBOARD_ADDRESS = "0x59ff2595588aa2236441b0e82b2cd692e1373e58"
    USDC_ADDRESS = "0x036CbD53842c5426634e7929541eC2318f3dCF7e"  # USDC on Base Sepolia
    
    def __init__(self, private_key: Optional[str] = None, rpc_url: Optional[str] = None):
        """
        Initialize Web3 client
        
        Args:
            private_key: Player's private key (optional for read-only operations)
            rpc_url: Custom RPC URL (optional, defaults to Base Sepolia)
        """
        self.rpc_url = rpc_url or self.BASE_SEPOLIA_RPC
        self.w3 = Web3(Web3.HTTPProvider(self.rpc_url))
        
        if not self.w3.is_connected():
            raise ConnectionError(f"Failed to connect to {self.rpc_url}")
        
        # Set up account if private key provided
        self.account = None
        if private_key:
            self.account = Account.from_key(private_key)
            self.w3.eth.default_account = self.account.address
        
        # Load contract ABIs and initialize contracts
        self._load_contracts()
    
    def _load_contracts(self):
        """Load contract ABIs and initialize contract instances"""
        # Note: In production, you'd load these from files
        # For now, we'll define minimal ABIs for the functions we need
        
        self.game_economy_abi = [
            {
                "inputs": [],
                "name": "purchaseHealthWithETH",
                "outputs": [],
                "stateMutability": "payable",
                "type": "function"
            },
            {
                "inputs": [{"name": "_amount", "type": "uint256"}],
                "name": "purchaseHealthWithUSDC",
                "outputs": [],
                "stateMutability": "nonpayable",
                "type": "function"
            },
            {
                "inputs": [],
                "name": "getCurrentETHPrice",
                "outputs": [{"name": "", "type": "uint256"}],
                "stateMutability": "view",
                "type": "function"
            },
            {
                "inputs": [],
                "name": "getCurrentUSDCPrice",
                "outputs": [{"name": "", "type": "uint256"}],
                "stateMutability": "view",
                "type": "function"
            },
            {
                "inputs": [
                    {"name": "_player", "type": "address"},
                    {"name": "_playerName", "type": "string"},
                    {"name": "_score", "type": "uint256"},
                    {"name": "_healthPurchased", "type": "bool"}
                ],
                "name": "endGameSession",
                "outputs": [],
                "stateMutability": "nonpayable",
                "type": "function"
            }
        ]
        
        self.leaderboard_abi = [
            {
                "inputs": [],
                "name": "getAllTimeLeaderboard",
                "outputs": [
                    {
                        "components": [
                            {"name": "player", "type": "address"},
                            {"name": "playerName", "type": "string"},
                            {"name": "score", "type": "uint256"},
                            {"name": "timestamp", "type": "uint256"},
                            {"name": "isPaidPlayer", "type": "bool"}
                        ],
                        "name": "",
                        "type": "tuple[]"
                    }
                ],
                "stateMutability": "view",
                "type": "function"
            },
            {
                "inputs": [],
                "name": "getDailyLeaderboard",
                "outputs": [
                    {
                        "components": [
                            {"name": "player", "type": "address"},
                            {"name": "playerName", "type": "string"},
                            {"name": "score", "type": "uint256"},
                            {"name": "timestamp", "type": "uint256"},
                            {"name": "isPaidPlayer", "type": "bool"}
                        ],
                        "name": "",
                        "type": "tuple[]"
                    }
                ],
                "stateMutability": "view",
                "type": "function"
            },
            {
                "inputs": [{"name": "_player", "type": "address"}],
                "name": "playerStats",
                "outputs": [
                    {"name": "bestScore", "type": "uint256"},
                    {"name": "totalGames", "type": "uint256"},
                    {"name": "totalSpent", "type": "uint256"},
                    {"name": "lastKnownName", "type": "string"}
                ],
                "stateMutability": "view",
                "type": "function"
            }
        ]
        
        self.usdc_abi = [
            {
                "inputs": [
                    {"name": "spender", "type": "address"},
                    {"name": "amount", "type": "uint256"}
                ],
                "name": "approve",
                "outputs": [{"name": "", "type": "bool"}],
                "stateMutability": "nonpayable",
                "type": "function"
            },
            {
                "inputs": [{"name": "account", "type": "address"}],
                "name": "balanceOf",
                "outputs": [{"name": "", "type": "uint256"}],
                "stateMutability": "view",
                "type": "function"
            }
        ]
        
        # Initialize contract instances
        self.game_economy = self.w3.eth.contract(
            address=self.GAME_ECONOMY_ADDRESS,
            abi=self.game_economy_abi
        )
        
        self.leaderboard = self.w3.eth.contract(
            address=self.LEADERBOARD_ADDRESS,
            abi=self.leaderboard_abi
        )
        
        self.usdc = self.w3.eth.contract(
            address=self.USDC_ADDRESS,
            abi=self.usdc_abi
        )
    
    def get_health_prices(self) -> Dict[str, float]:
        """Get current health purchase prices in ETH and USDC"""
        try:
            eth_price_wei = self.game_economy.functions.getCurrentETHPrice().call()
            usdc_price_units = self.game_economy.functions.getCurrentUSDCPrice().call()
            
            return {
                "eth_price": self.w3.from_wei(eth_price_wei, 'ether'),
                "usdc_price": usdc_price_units / 1e6,  # USDC has 6 decimals
                "eth_price_wei": eth_price_wei,
                "usdc_price_units": usdc_price_units
            }
        except Exception as e:
            print(f"Error getting health prices: {e}")
            return {"eth_price": 0, "usdc_price": 0, "eth_price_wei": 0, "usdc_price_units": 0}
    
    def purchase_health_with_eth(self) -> Optional[str]:
        """Purchase health with ETH"""
        if not self.account:
            raise ValueError("Account required for transactions")
        
        try:
            # Get current price
            prices = self.get_health_prices()
            eth_price = prices["eth_price_wei"]
            
            if eth_price == 0:
                raise ValueError("Could not get ETH price")
            
            # Build transaction
            transaction = self.game_economy.functions.purchaseHealthWithETH().build_transaction({
                'from': self.account.address,
                'value': eth_price,
                'gas': 200000,
                'gasPrice': self.w3.eth.gas_price,
                'nonce': self.w3.eth.get_transaction_count(self.account.address),
                'chainId': self.CHAIN_ID
            })
            
            # Sign and send transaction
            signed_txn = self.w3.eth.account.sign_transaction(transaction, self.account.key)
            tx_hash = self.w3.eth.send_raw_transaction(signed_txn.rawTransaction)
            
            print(f"Health purchased with ETH! TX: {tx_hash.hex()}")
            return tx_hash.hex()
            
        except Exception as e:
            print(f"Error purchasing health with ETH: {e}")
            return None
    
    def purchase_health_with_usdc(self) -> Optional[str]:
        """Purchase health with USDC (requires prior approval)"""
        if not self.account:
            raise ValueError("Account required for transactions")
        
        try:
            # Get current price
            prices = self.get_health_prices()
            usdc_amount = prices["usdc_price_units"]
            
            if usdc_amount == 0:
                raise ValueError("Could not get USDC price")
            
            # Check USDC balance
            balance = self.usdc.functions.balanceOf(self.account.address).call()
            if balance < usdc_amount:
                raise ValueError(f"Insufficient USDC balance. Need {usdc_amount/1e6}, have {balance/1e6}")
            
            # Build transaction
            transaction = self.game_economy.functions.purchaseHealthWithUSDC(usdc_amount).build_transaction({
                'from': self.account.address,
                'gas': 250000,
                'gasPrice': self.w3.eth.gas_price,
                'nonce': self.w3.eth.get_transaction_count(self.account.address),
                'chainId': self.CHAIN_ID
            })
            
            # Sign and send transaction
            signed_txn = self.w3.eth.account.sign_transaction(transaction, self.account.key)
            tx_hash = self.w3.eth.send_raw_transaction(signed_txn.rawTransaction)
            
            print(f"Health purchased with USDC! TX: {tx_hash.hex()}")
            return tx_hash.hex()
            
        except Exception as e:
            print(f"Error purchasing health with USDC: {e}")
            return None
    
    def approve_usdc(self, amount: Optional[int] = None) -> Optional[str]:
        """Approve USDC spending for health purchases"""
        if not self.account:
            raise ValueError("Account required for transactions")
        
        try:
            # Use max approval if no amount specified
            if amount is None:
                amount = 2**256 - 1  # Max uint256
            
            # Build transaction
            transaction = self.usdc.functions.approve(self.GAME_ECONOMY_ADDRESS, amount).build_transaction({
                'from': self.account.address,
                'gas': 100000,
                'gasPrice': self.w3.eth.gas_price,
                'nonce': self.w3.eth.get_transaction_count(self.account.address),
                'chainId': self.CHAIN_ID
            })
            
            # Sign and send transaction
            signed_txn = self.w3.eth.account.sign_transaction(transaction, self.account.key)
            tx_hash = self.w3.eth.send_raw_transaction(signed_txn.rawTransaction)
            
            print(f"USDC approved! TX: {tx_hash.hex()}")
            return tx_hash.hex()
            
        except Exception as e:
            print(f"Error approving USDC: {e}")
            return None
    
    def submit_game_score(self, player_name: str, score: int, health_purchased: bool) -> Optional[str]:
        """Submit game score to leaderboard"""
        if not self.account:
            raise ValueError("Account required for transactions")
        
        try:
            # Build transaction
            transaction = self.game_economy.functions.endGameSession(
                self.account.address,
                player_name,
                score,
                health_purchased
            ).build_transaction({
                'from': self.account.address,
                'gas': 300000,
                'gasPrice': self.w3.eth.gas_price,
                'nonce': self.w3.eth.get_transaction_count(self.account.address),
                'chainId': self.CHAIN_ID
            })
            
            # Sign and send transaction
            signed_txn = self.w3.eth.account.sign_transaction(transaction, self.account.key)
            tx_hash = self.w3.eth.send_raw_transaction(signed_txn.rawTransaction)
            
            print(f"Score submitted! TX: {tx_hash.hex()}")
            return tx_hash.hex()
            
        except Exception as e:
            print(f"Error submitting score: {e}")
            return None
    
    def get_leaderboard(self, leaderboard_type: str = "all_time") -> List[Dict]:
        """Get leaderboard data"""
        try:
            if leaderboard_type == "daily":
                entries = self.leaderboard.functions.getDailyLeaderboard().call()
            else:
                entries = self.leaderboard.functions.getAllTimeLeaderboard().call()
            
            # Convert to readable format
            leaderboard = []
            for entry in entries:
                leaderboard.append({
                    "player": entry[0],
                    "name": entry[1],
                    "score": entry[2],
                    "timestamp": datetime.fromtimestamp(entry[3]).strftime("%Y-%m-%d %H:%M:%S"),
                    "is_paid_player": entry[4]
                })
            
            return leaderboard
            
        except Exception as e:
            print(f"Error getting leaderboard: {e}")
            return []
    
    def get_player_stats(self, player_address: Optional[str] = None) -> Dict:
        """Get player statistics"""
        if not player_address and not self.account:
            raise ValueError("Player address or account required")
        
        address = player_address or self.account.address
        
        try:
            stats = self.leaderboard.functions.playerStats(address).call()
            return {
                "best_score": stats[0],
                "total_games": stats[1],
                "total_spent": stats[2] / 1e6,  # Convert from USDC units
                "last_known_name": stats[3]
            }
        except Exception as e:
            print(f"Error getting player stats: {e}")
            return {"best_score": 0, "total_games": 0, "total_spent": 0, "last_known_name": ""}
    
    def get_account_info(self) -> Dict:
        """Get current account information"""
        if not self.account:
            return {"address": None, "eth_balance": 0, "usdc_balance": 0}
        
        try:
            eth_balance = self.w3.eth.get_balance(self.account.address)
            usdc_balance = self.usdc.functions.balanceOf(self.account.address).call()
            
            return {
                "address": self.account.address,
                "eth_balance": self.w3.from_wei(eth_balance, 'ether'),
                "usdc_balance": usdc_balance / 1e6
            }
        except Exception as e:
            print(f"Error getting account info: {e}")
            return {"address": self.account.address, "eth_balance": 0, "usdc_balance": 0}


# Example usage and testing
if __name__ == "__main__":
    # Example usage (replace with actual private key for testing)
    # client = SaveTheCastleWeb3Client(private_key="your_private_key_here")
    
    # For read-only operations, no private key needed
    client = SaveTheCastleWeb3Client()
    
    print("Save the Castle Web3 Client initialized")
    print(f"Connected to: {client.rpc_url}")
    print(f"GameEconomy: {client.GAME_ECONOMY_ADDRESS}")
    print(f"Leaderboard: {client.LEADERBOARD_ADDRESS}")
    
    # Test read operations
    print("\n--- Current Health Prices ---")
    prices = client.get_health_prices()
    print(f"ETH Price: {prices['eth_price']} ETH")
    print(f"USDC Price: {prices['usdc_price']} USDC")
    
    print("\n--- All-Time Leaderboard ---")
    leaderboard = client.get_leaderboard("all_time")
    for i, entry in enumerate(leaderboard[:5]):  # Top 5
        print(f"{i+1}. {entry['name']} - {entry['score']} points ({entry['timestamp']})")

File: web-app/src/blockchain/Web3Client.ts
import { ethers, BrowserProvider, Contract } from 'ethers';

export interface LeaderboardEntry {
  player: string;
  name: string;
  score: number;
  timestamp: string;
  is_paid_player: boolean;
}

export interface ScoreData {
  score: number;
  healthPurchased: boolean;
  gameTime: number;
}

export interface HealthPrices {
  ethPrice: number;
  usdcPrice: number;
  ethPriceWei: bigint;
  usdcPriceUnits: bigint;
}

export class Web3Client {
  private provider: BrowserProvider | null = null;
  private signer: ethers.Signer | null = null;
  private gameEconomyContract: Contract | null = null;
  private leaderboardContract: Contract | null = null;
  private usdcContract: Contract | null = null;

  // Base Sepolia configuration
  private readonly CHAIN_ID = 84532;
  private readonly RPC_URL = 'https://sepolia.base.org';
  
  // Contract addresses (checksummed)
  private readonly GAME_ECONOMY_ADDRESS = '0x55cBEa71ad8B981B91B137116B76a4828F90C548';
  private readonly LEADERBOARD_ADDRESS = '0x59FF2595588AA2236441B0E82b2CD692e1373E58';
  private readonly USDC_ADDRESS = '0x036CbD53842c5426634e7929541eC2318f3dCF7e';

  // Minimal ABIs for the functions we need
  private readonly GAME_ECONOMY_ABI = [
    'function purchaseHealthWithETH() external payable',
    'function purchaseHealthWithUSDC(uint256 _amount) external',
    'function getCurrentETHPrice() external view returns (uint256)',
    'function getCurrentUSDCPrice() external view returns (uint256)',
    'function endGameSession(address _player, string _playerName, uint256 _score, bool _healthPurchased) external'
  ];

  private readonly LEADERBOARD_ABI = [
    'function getAllTimeLeaderboard() external view returns (tuple(address player, string playerName, uint256 score, uint256 timestamp, bool isPaidPlayer)[])',
    'function getDailyLeaderboard() external view returns (tuple(address player, string playerName, uint256 score, uint256 timestamp, bool isPaidPlayer)[])',
    'function playerStats(address _player) external view returns (uint256 bestScore, uint256 totalGames, uint256 totalSpent, string lastKnownName)'
  ];

  private readonly USDC_ABI = [
    'function approve(address spender, uint256 amount) external returns (bool)',
    'function balanceOf(address account) external view returns (uint256)'
  ];

  async connectWallet(): Promise<{ success: boolean; address?: string; error?: string }> {
    try {
      if (!window.ethereum) {
        return { success: false, error: 'No wallet found. Please install MetaMask or another Web3 wallet.' };
      }

      // Request account access
      await window.ethereum.request({ method: 'eth_requestAccounts' });
      
      this.provider = new BrowserProvider(window.ethereum);
      this.signer = await this.provider.getSigner();
      
      const address = await this.signer.getAddress();
      
      // Initialize contracts
      this.initializeContracts();
      
      // Check/switch to Base Sepolia
      await this.switchToBaseSepolia();
      
      return { success: true, address };
    } catch (error: any) {
      console.error('Wallet connection error:', error);
      return { success: false, error: error.message || 'Failed to connect wallet' };
    }
  }

  private initializeContracts(): void {
    if (!this.provider || !this.signer) return;

    this.gameEconomyContract = new Contract(
      this.GAME_ECONOMY_ADDRESS,
      this.GAME_ECONOMY_ABI,
      this.signer
    );

    this.leaderboardContract = new Contract(
      this.LEADERBOARD_ADDRESS,
      this.LEADERBOARD_ABI,
      this.signer
    );

    this.usdcContract = new Contract(
      this.USDC_ADDRESS,
      this.USDC_ABI,
      this.signer
    );
  }

  private async switchToBaseSepolia(): Promise<void> {
    if (!window.ethereum) return;

    try {
      await window.ethereum.request({
        method: 'wallet_switchEthereumChain',
        params: [{ chainId: `0x${this.CHAIN_ID.toString(16)}` }],
      });
    } catch (switchError: any) {
      // Chain not added, try to add it
      if (switchError.code === 4902) {
        await window.ethereum.request({
          method: 'wallet_addEthereumChain',
          params: [{
            chainId: `0x${this.CHAIN_ID.toString(16)}`,
            chainName: 'Base Sepolia',
            nativeCurrency: {
              name: 'ETH',
              symbol: 'ETH',
              decimals: 18,
            },
            rpcUrls: [this.RPC_URL],
            blockExplorerUrls: ['https://sepolia.basescan.org/'],
          }],
        });
      }
    }
  }

  async getHealthPrices(): Promise<HealthPrices | null> {
    if (!this.gameEconomyContract) {
      console.warn('Contract not initialized');
      return null;
    }

    try {
      const [ethPriceWei, usdcPriceUnits] = await Promise.all([
        this.gameEconomyContract.getCurrentETHPrice(),
        this.gameEconomyContract.getCurrentUSDCPrice()
      ]);

      return {
        ethPrice: parseFloat(ethers.formatEther(ethPriceWei)),
        usdcPrice: parseFloat(ethers.formatUnits(usdcPriceUnits, 6)),
        ethPriceWei: BigInt(ethPriceWei.toString()),
        usdcPriceUnits: BigInt(usdcPriceUnits.toString())
      };
    } catch (error) {
      console.error('Error getting health prices:', error);
      return null;
    }
  }

  async purchaseHealthWithETH(): Promise<{ success: boolean; txHash?: string; error?: string }> {
    if (!this.gameEconomyContract) {
      return { success: false, error: 'Contract not initialized' };
    }

    try {
      const prices = await this.getHealthPrices();
      if (!prices) {
        return { success: false, error: 'Could not get current prices' };
      }

      const tx = await this.gameEconomyContract.purchaseHealthWithETH({
        value: prices.ethPriceWei
      });

      await tx.wait();
      return { success: true, txHash: tx.hash };
    } catch (error: any) {
      console.error('ETH purchase error:', error);
      return { success: false, error: error.message || 'Purchase failed' };
    }
  }

  async purchaseHealthWithUSDC(): Promise<{ success: boolean; txHash?: string; error?: string }> {
    if (!this.gameEconomyContract || !this.usdcContract) {
      return { success: false, error: 'Contracts not initialized' };
    }

    try {
      const prices = await this.getHealthPrices();
      if (!prices) {
        return { success: false, error: 'Could not get current prices' };
      }

      // Check USDC balance
      const address = await this.signer!.getAddress();
      const balance = await this.usdcContract.balanceOf(address);
      
      if (balance < prices.usdcPriceUnits) {
        return { 
          success: false, 
          error: `Insufficient USDC balance. Need ${prices.usdcPrice} USDC` 
        };
      }

      const tx = await this.gameEconomyContract.purchaseHealthWithUSDC(prices.usdcPriceUnits);
      await tx.wait();
      return { success: true, txHash: tx.hash };
    } catch (error: any) {
      console.error('USDC purchase error:', error);
      return { success: false, error: error.message || 'Purchase failed' };
    }
  }

  async approveUSDC(amount?: bigint): Promise<{ success: boolean; txHash?: string; error?: string }> {
    if (!this.usdcContract) {
      return { success: false, error: 'USDC contract not initialized' };
    }

    try {
      const approveAmount = amount || ethers.MaxUint256;
      const tx = await this.usdcContract.approve(this.GAME_ECONOMY_ADDRESS, approveAmount);
      await tx.wait();
      return { success: true, txHash: tx.hash };
    } catch (error: any) {
      console.error('USDC approval error:', error);
      return { success: false, error: error.message || 'Approval failed' };
    }
  }

  async submitScore(playerName: string, scoreData: ScoreData): Promise<{ success: boolean; txHash?: string; error?: string }> {
    if (!this.gameEconomyContract) {
      return { success: false, error: 'Contract not initialized' };
    }

    try {
      const address = await this.signer!.getAddress();
      
      const tx = await this.gameEconomyContract.endGameSession(
        address,
        playerName,
        scoreData.score,
        scoreData.healthPurchased
      );

      await tx.wait();
      return { success: true, txHash: tx.hash };
    } catch (error: any) {
      console.error('Score submission error:', error);
      return { success: false, error: error.message || 'Score submission failed' };
    }
  }

  async getLeaderboard(type: 'all_time' | 'daily' = 'all_time'): Promise<LeaderboardEntry[]> {
    if (!this.leaderboardContract) {
      console.warn('Leaderboard contract not initialized');
      return [];
    }

    try {
      const entries = type === 'daily' 
        ? await this.leaderboardContract.getDailyLeaderboard()
        : await this.leaderboardContract.getAllTimeLeaderboard();

      return entries.map((entry: any) => ({
        player: entry.player,
        name: entry.playerName,
        score: parseInt(entry.score.toString()),
        timestamp: new Date(parseInt(entry.timestamp.toString()) * 1000).toISOString(),
        is_paid_player: entry.isPaidPlayer
      }));
    } catch (error) {
      console.error('Error getting leaderboard:', error);
      return [];
    }
  }

  async getPlayerStats(playerAddress?: string): Promise<any> {
    if (!this.leaderboardContract) {
      return null;
    }

    try {
      const address = playerAddress || await this.signer!.getAddress();
      const stats = await this.leaderboardContract.playerStats(address);
      
      return {
        bestScore: parseInt(stats.bestScore.toString()),
        totalGames: parseInt(stats.totalGames.toString()),
        totalSpent: parseFloat(ethers.formatUnits(stats.totalSpent, 6)),
        lastKnownName: stats.lastKnownName
      };
    } catch (error) {
      console.error('Error getting player stats:', error);
      return null;
    }
  }

  async getAccountInfo(): Promise<{ address: string; ethBalance: number; usdcBalance: number } | null> {
    if (!this.signer || !this.provider) return null;

    try {
      const address = await this.signer.getAddress();
      const ethBalance = await this.provider.getBalance(address);
      
      let usdcBalance = BigInt(0);
      if (this.usdcContract) {
        usdcBalance = await this.usdcContract.balanceOf(address);
      }

      return {
        address,
        ethBalance: parseFloat(ethers.formatEther(ethBalance)),
        usdcBalance: parseFloat(ethers.formatUnits(usdcBalance, 6))
      };
    } catch (error) {
      console.error('Error getting account info:', error);
      return null;
    }
  }

  isConnected(): boolean {
    return this.provider !== null && this.signer !== null;
  }

  getAddress(): Promise<string> | null {
    return this.signer?.getAddress() || null;
  }
}

// Global Web3Client instance
export const web3Client = new Web3Client();